pkg/cmd/roachtest/quit.go

```diff
@@ -0,0 +1,226 @@
+// Copyright 2020 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package main
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
+	"github.com/cockroachdb/cockroach/pkg/util/contextutil"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/errors"
+)
+
+// runQuitTransfersLeases performs rolling restarts on a
+// 3-node cluster and ascertains that each node shutting down
+// transfers all its leases reliably to other nodes prior to
+// terminating.
+func runQuitTransfersLeases(
+	ctx context.Context,
+	t *test,
+	c *cluster,
+	methodName string,
+	method func(ctx context.Context, t *test, c *cluster, nodeID int),
+) {
+	args := startArgs("--env=COCKROACH_SCAN_MAX_IDLE_TIME=5ms")
+	c.Put(ctx, cockroach, "./cockroach")
+	c.Start(ctx, t, args)
+
+	// rwt runs a command with a 1-minute timeout.
+	rwt := func(fn func(ctx context.Context)) {
+		if err := contextutil.RunWithTimeout(ctx, "do", time.Minute, func(ctx context.Context) error {
+			fn(ctx)
+			return nil
+		}); err != nil {
+			t.Fatal(err)
+		}
+	}
+
+	// Now start the remainder of the nodes, then wait for
+	// up-replication.
+	func() {
+		db := c.Conn(ctx, 1)
+		defer db.Close()
+
+		// We'll want rebalancing to be a bit fast than normal, so
+		// that the up-replication does not take ages.
+		if _, err := db.ExecContext(ctx, `SET CLUSTER SETTING	kv.snapshot_rebalance.max_rate = '128MiB'`); err != nil {
+			t.Fatal(err)
+		}
+
+		err := retry.ForDuration(30*time.Second, func() error {
+			t.l.Printf("waiting for up-replication\n")
+			row := db.QueryRowContext(ctx, `SELECT min(array_length(replicas, 1)) FROM crdb_internal.ranges_no_leases`)
+			minReplicas := 0
+			if err := row.Scan(&minReplicas); err != nil {
+				t.Fatal(err)
+			}
+			if minReplicas < 3 {
+				time.Sleep(time.Second)
+				return errors.Newf("some ranges not up-replicated yet")
+			}
+			return nil
+		})
+		if err != nil {
+			t.Fatalf("cluster did not up-replicate: %v", err)
+		}
+	}()
+
+	// Create a bunch of ranges.
+	const numRanges = 500
+	func() {
+		db := c.Conn(ctx, 1)
+		defer db.Close()
+		if _, err := db.ExecContext(ctx, fmt.Sprintf(`
+CREATE TABLE t(x, y, PRIMARY KEY(x)) AS SELECT @1, 1 FROM generate_series(1,%[1]d)`,
+			numRanges)); err != nil {
+			t.Fatal(err)
+		}
+		// We split them from right-to-left so we're peeling at most 1
+		// row each time on the right.
+		for i := numRanges; i > 1; i -= 100 {
+			t.l.Printf("creating 100 ranges (%d-%d)...\n", i, i-99)
```

Comment:
nit: this seems more complicated than I would've imagined. Why not a single `ALTER TABLE SPLIT AT`? Aren't you doing only five laps through this loop?